### PR TITLE
Implement custom sidebar messages, resolves Electroid#157

### PR DESCRIPTION
**Pull request by Meeples10**
Signed-off-by: Meeples10 <8867705+Meeples10@users.noreply.github.com>

### CODE REVIEW

File: src/main/resources/config.yml [hunk 1]
```diff
@@ -108,4 +108,11 @@ experiments:
...
-  destroy-match-delay: 100
\ No newline at end of file
+  destroy-match-delay: 100
+
+sidebar-message:
+  enabled: false
+  # Text may be formatted with '&'
+  text: ""
+  # If the message should be positioned at the top or bottom of the sidebar
+  top: false
```
Comment (TheMolkaPL) [inline src/main/resources/config.yml]: Missing new line

File: src/main/java/tc/oc/pgm/Config.java [hunk 2]
```diff
@@ -384,4 +384,19 @@ public int getDestroyMatchDelay() {
...
+    }
+
+    public static String format() {
+      return ChatColor.translateAlternateColorCodes(
+          '&', getConfiguration().getString("sidebar-message.text"));
+    }
+
+    public static boolean atTop() {
+      return getConfiguration().getBoolean("sidebar-message.top", false);
+    }
```
Comment (Pablete1234) [inline src/main/java/tc/oc/pgm/Config.java]: Would maybe be easier to have message-top and message-bottom, with the ability to leave either of them blank?
That way you convert 3 settings into 2 (remove enabled & atTop, keep message-top and message-bottom), it gives more flexibility for people trying to configure it, and it's (maybe?) easier to code the condition.

File: src/main/java/tc/oc/pgm/scoreboard/SidebarMatchModule.java [hunk 3]
```diff
@@ -463,6 +463,17 @@ public int compare(Goal a, Goal b) {
...
+        if (Config.SidebarMessage.atTop()) {
+          rows.add(0, Config.SidebarMessage.format());
+        } else {
+          if (rows.size() >= MAX_ROWS) {
+            rows.set(MAX_ROWS - 1, Config.SidebarMessage.format());
+          } else {
+            rows.add(Config.SidebarMessage.format());
+          }
+        }
+      }
```
Comment (Pablete1234) [inline src/main/java/tc/oc/pgm/scoreboard/SidebarMatchModule.java]: Can we have a player setting to disable scoreboard messages?

Comment (TheMolkaPL) [inline src/main/java/tc/oc/pgm/scoreboard/SidebarMatchModule.java]: I think there should a setting `overwrite-existing` default to `false`. If they scoreboard if full already, we can lose important information about the objectives/match.

Comment (Meeples10) [reply src/main/java/tc/oc/pgm/scoreboard/SidebarMatchModule.java]: This wouldn't be possible with the way sidebars are currently rendered. The sidebar doesn't know which player is viewing it, only the player's team.

Comment (TheMolkaPL) [reply src/main/java/tc/oc/pgm/scoreboard/SidebarMatchModule.java]: Don't add the setting then lol ;p
https://github.com/Electroid/PGM/pull/176/commits/98246c32928f7490406181d41dd0c4a0d7004323#diff-f08f8cdb20626aaae1393d6a6ea2a79fR13-R16

**Comment by TheMolkaPL** [comment]
Probably because it would require refactor of the whole sidebar rendering logic.

File: src/main/resources/config.yml [hunk 4]
```diff
@@ -108,4 +108,12 @@ experiments:
   # Defer tab rendering to group more updates at once, in ticks
   tab-render-delay: 5
   # Ticks to wait to destroy older match
-  destroy-match-delay: 100
\ No newline at end of file
+  destroy-match-delay: 100
+
+sidebar-message:
```
Comment (Electroid) [inline src/main/resources/config.yml]: `sidebar`

File: src/main/resources/config.yml [hunk 5]
```diff
@@ -108,4 +108,12 @@ experiments:
...
+  # Text may be formatted with '&'
+  # Leave blank to disable
+  message-top: ""
```
Comment (Electroid) [inline src/main/resources/config.yml]: `top`

File: src/main/resources/config.yml [hunk 6]
```diff
@@ -108,4 +108,12 @@ experiments:
...
+  message-bottom: ""
```
Comment (Electroid) [inline src/main/resources/config.yml]: `bottom`

File: src/main/resources/config.yml [hunk 7]
```diff
@@ -108,4 +108,12 @@ experiments:
...
+  # If the sidebar message should overwrite lines in the sidebar when it is full
+  overwrite-existing: false
```
Comment (Electroid) [inline src/main/resources/config.yml]: `overwrite`

**Pull request merged**